android-vjestina-homework-movie-app-data-mock/app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt

```diff
@@ -0,0 +1,125 @@
+                    MovieCategory.POPULAR_ON_TV -> movieService.fetchTopRatedMovies()
+                    MovieCategory.POPULAR_FOR_RENT -> movieService.fetchNowPlayingMovies()
+                    MovieCategory.POPULAR_IN_THEATERS -> movieService.fetchUpcomingMovies()
+                    MovieCategory.UPCOMING_TODAY -> movieService.fetchUpcomingMovies()
+                    MovieCategory.UPCOMING_THIS_WEEK -> movieService.fetchPopularMovies()
+                }
+                emit(movieResponse.movies)
+            }.flatMapLatest { apiMovies ->
+                movieDao.favorites().map { favoriteMovies ->
+                    apiMovies.map { apiMovie ->
+                        apiMovie.toMovie(isFavorite = favoriteMovies.any { it.id == apiMovie.id })
+                    }
+                }
+            }.shareIn(
+                scope = CoroutineScope(bgDispatcher),
+                started = SharingStarted.WhileSubscribed(1000L),
+                replay = 1,
+            )
+        }
+
+    private val favorites = movieDao.favorites().map {
+        it.map { dbFavoriteMovie ->
+            Movie(
+                id = dbFavoriteMovie.id,
+                imageUrl = dbFavoriteMovie.posterUrl,
+                title = "",
+                overview = "",
+                isFavorite = true,
+            )
+        }
+    }.shareIn(
+        scope = CoroutineScope(bgDispatcher),
+        started = SharingStarted.WhileSubscribed(1000L),
+        replay = 1,
+    )
+
+    override fun movies(movieCategory: MovieCategory): Flow<List<Movie>> =
+        moviesByCategory[movieCategory]!!
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    override fun movieDetails(movieId: Int): Flow<MovieDetails> = flow {
+        emit(movieService.fetchMovieDetails(movieId) to movieService.fetchMovieCredits(movieId))
+    }.flatMapLatest { (apiMovieDetails, apiMovieCredits) ->
+        movieDao.favorites().map { favoriteMovies ->
+            apiMovieDetails.toMovieDetails(
+                isFavorite = favoriteMovies.any { it.id == apiMovieDetails.id },
+                crew = apiMovieCredits.crew,
+                cast = apiMovieCredits.cast
+            )
+        }
+    }.flowOn(bgDispatcher)
+
+    override fun favoriteMovies(): Flow<List<Movie>> = favorites
+
+    override suspend fun addMovieToFavorites(movieId: Int) {
+        runBlocking(bgDispatcher) {
+            movieDao.insertMovie(
+                DbFavoriteMovie(
+                    movieId,
+                    "$BASE_IMAGE_URL/${movieService.fetchMovieDetails(movieId).poster_path}"
+                )
+            )
+        }
+    }
+
+    private suspend fun findMovie(movieId: Int): Movie? {
+        var movie: Movie? = null
+        moviesByCategory.values.forEach { value ->
+            val movies = value.first()
+            movies.forEach {
+                if (it.id == movieId) {
+                    movie = it
```

Comment:
instead of saving it to movie and then returning at the end you can return to movie here.